ee/libc/src/sleep.c

```diff
@@ -14,83 +14,23 @@
  */
 
 #include <kernel.h>
-
+#include <unistd.h>
 #include <errno.h>
 #include <time.h>
-#include <unistd.h>
-
-#include "ps2sdkapi.h"
-
-#define MIN_HSYNC_PER_SEC (240*50)
-#define MIN_HSYNC_DELAY 100
-#define MAX_PS2CLOCK_PER_HSYNC (PS2_CLOCKS_PER_SEC/MIN_HSYNC_PER_SEC)
+#include <timer_alarm.h>
 
 #ifdef F_nanosleep
-// Start with the highest possible value (so we sleep too short)
-static unsigned int iPS2ClockPerHSync = MAX_PS2CLOCK_PER_HSYNC;
-
-static void _sleep_waker(s32 alarm_id, u16 time, void *arg2)
-{
-    s32 *pSema = (s32 *)arg2;
-
-    (void)alarm_id;
-    (void)time;
-
-    iSignalSema(*pSema);
-    ExitHandler();
-}
-
-static inline u64 timespec_to_us(const struct timespec *pts) {
-    return (pts->tv_sec * 1000000ULL) + (pts->tv_nsec / 1000ULL);
-}
-
-static inline ps2_clock_t us_to_ps2_clock(u64 us) {
-    return (us * PS2_CLOCKS_PER_MSEC) / 1000;
-}
-
-static inline ps2_clock_t timespec_to_ps2_clock(const struct timespec *pts) {
-    return us_to_ps2_clock(timespec_to_us(pts));
+static inline u64 timespec_to_nanosecs(const struct timespec *pts) {
+    return (pts->tv_sec * 1000000000ULL) + pts->tv_nsec;
 }
 
 int nanosleep(const struct timespec *req, struct timespec *rem)
 {
-    ee_sema_t sema;
-    s32 sema_id;
-    ps2_clock_t clock_delay, clock_end, clock_real_end;
-    u16 hsync_delay;
-
-    clock_delay = timespec_to_ps2_clock(req);
-    clock_end   = ps2_clock() + clock_delay;
-    hsync_delay = clock_delay / iPS2ClockPerHSync;
+    u64 nano_secs, cycles;
 
-    if (hsync_delay >= MIN_HSYNC_DELAY) {
-        sema.init_count = 0;
-        sema.max_count  = 1;
-        sema.option     = 0;
-        sema_id = CreateSema(&sema);
-
-        SetAlarm(hsync_delay, _sleep_waker, &sema_id);
-        WaitSema(sema_id);
-        DeleteSema(sema_id);
-
-        // Correct the HSync delay
-        // NOTE: We're always trying to end btween 0 and 1 HSyncs too short
-        clock_real_end = ps2_clock();
-        if (clock_real_end < (clock_end - iPS2ClockPerHSync)) {
-            // We sleeped too short, so HSyncs must be faster
-            // Correct the HSyncs for next time
-            iPS2ClockPerHSync -= (clock_end - iPS2ClockPerHSync - clock_real_end) / hsync_delay;
-        }
-        else if (clock_real_end > clock_end) {
-            // We sleeped too long, so HSyncs must be slower
-            // Correct the HSyncs for next time
-            iPS2ClockPerHSync += (clock_real_end - clock_end) / (hsync_delay-1);
-        }
-    }
-
-    // Delay whatever time is left
-    while(clock_end > ps2_clock())
-        ;
+    nano_secs = timespec_to_nanosecs(req);
+    cycles = NSecToCycles(nano_secs);
```

Comment:
Have the official TimerUSec2BusClock / TimerBusClock2USec functions also been implemented?

Reply:
Gave them a different name. We can rename them if you prefer.
Also the cycles to time functions are not implemented. I didnt need to use them, but perhaps they do have a use for some users?

Reply:
I have added a new commit where I have implemented them, and also I have changed the name of the other to be "closer" to the PS2 official ones